feat: beta composition schema aware validation (#576)
crossplane/docs: content/master/concepts/compositions.md

* Composition function `type` is `container`.
* Composition function names are unique.

### Composition schema aware validation

On top of the preceding integrity checks, Crossplane performs schema aware
validation of Compositions, checking that `patches`, `readinessChecks` and
`connectionDetails` are valid according to the involved resources' schemas, for
example checking that the source/destination field of a patch is a valid field
according to the source/destination resource's schema and that types match, if
defined, taking into account transforms.

{{<hint "note" >}}

> Composition schema aware validation is a beta feature, so Crossplane
> enables it by default. You can disable it by setting the
> `--enable-composition-webhook-schema-validation=false` flag on the Crossplane
> pod.
> The [Crossplane Pods]({{<ref "./pods#edit-the-deployment">}}) page has
> more information on enabling Crossplane flags.

Reviewer: ```suggestion
Composition schema aware validation is a beta feature. Crossplane enables
beta features by default. 

Disable schema aware validation by setting the
`--enable-composition-webhook-schema-validation=false` flag on the Crossplane
pod.  

The [Crossplane Pods]({{<ref "./pods#edit-the-deployment">}}) page has
more information on enabling Crossplane flags.
```

{{< /hint >}}

#### Schema aware validation modes

> Integrity validation errors always result in a rejection of the Composition.
> 
> You can configure how Crossplane should handle both missing involved resources'
> schemas and actual schema aware validation errors.

Reviewer: ```suggestion
<!-- vale alex.ProfanityUnlikely = NO -->
Integrity validation errors reject a Composition.
<!-- vale alex.ProfanityUnlikely = Yes  -->

Set the schema validation mode to configure how Crossplane handles both missing 
resource schemas and schema validation errors.
```

The word "reject" is a vale error and shouldn't be but I can't make the change in review so I just disable the check.

Author: I get the intention, but I feel it's slightly wrong saying that errors reject the composition... it's crossplane rejecting it because of errors. I'd rather be correct than have vale somewhat happy 😬

Reviewer: I'm trying to get away from "result in a rejection". How about 

"Crossplane rejects Compositions with validation errors"

?

The vale error from an imported library that says the word "reject" may be profane, which it isn't in our writing/use.

Author: ```suggestion
Crossplane always rejects Compositions in case of integrity errors.

Set the schema validation mode to configure how Crossplane handles both missing 
resource schemas and schema aware validation errors.
```

Reviewer: LGTM

{{<hint "note" >}}
In case of any involved resource schema missing, Crossplane skips schema aware

> validation altogether.

Reviewer: This is based on the validation mode according to the table below, right?

Author: That's for all modes, schema aware validation is not performed unless we have all required schemas available. Depending on the mode this could result in a warning or error.

{{< /hint >}}

The following modes are available:

> | Mode     | Missing Schema | Schema Aware Error | Integrity Error |
> | -------- | -------------- |--------------------|-----------------|
> | `warn`   | Warning        | Warning            | Error           |
> | `loose`  | Warning        | Error              | Error           |
> | `strict` | Error          | Error              | Error           |

Reviewer: ```suggestion
{{< table "table table-sm table-striped" >}}
| Mode     | Missing Schema | Schema Aware Error | Integrity Error |
| -------- | -------------- |--------------------|-----------------|
| `warn`   | Warning        | Warning            | Error           |
| `loose`  | Warning        | Error              | Error           |
| `strict` | Error          | Error              | Error           |
{{< /table >}}
```

table styling.

Change the validation mode for a Composition with the
{{<hover label="mode" line="5">}}crossplane.io/composition-schema-aware-validation-mode{{</hover>}} 
annotation.

If not specified, the default mode is `warn`.

> For example, to enable `loose` mode checking:

Reviewer: ```suggestion
For example, to enable `loose` mode checking set the annotation value to 
{{<hover label="mode" line="5">}}loose{{</hover>}}.
```

```yaml {copy-lines="none",label="mode"}
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  annotations:
    crossplane.io/composition-schema-aware-validation-mode: loose
  # Removed for brevity
spec:
  # Removed for brevity
```

{{<hint "important" >}}
Modes also apply to Compositions defined by Configuration packages, look at

> Crossplane's logs for warnings in case of issues. The default is just to warn,

Reviewer: Do you mean `kubectl logs` or the messages in `kubectl describe` of something?

Author: Logs logs, as in kubectl logs, describe ones are events, we log warnings to stdout if we get any back for one of the calls Crossplane does. E.g. the infamous "ControllerConfig has been deprecated" which showed up in the logs

* Composition function `type` is `container`.
* Composition function names are unique.

### Composition schema aware validation

Crossplane also performs schema aware
validation of Compositions. Schema validation checks that `patches`, 
`readinessChecks` and `connectionDetails` are valid according to the resource 
schemas. For example, checking that the source and destination fields of a patch 
are valid according to the source and destination resource schema.

{{<hint "note" >}}
Composition schema aware validation is a beta feature. Crossplane enables
beta features by default. 

Disable schema aware validation by setting the
`--enable-composition-webhook-schema-validation=false` flag on the Crossplane
pod.  

The [Crossplane Pods]({{<ref "./pods#edit-the-deployment">}}) page has
more information on enabling Crossplane flags.
{{< /hint >}}

#### Schema aware validation modes

Integrity validation errors always result in a rejection of the Composition.

You can configure how Crossplane should handle both missing involved resources'
schemas and actual schema aware validation errors.

{{<hint "note" >}}

> In case of any involved resource schema missing, Crossplane skips schema aware
> validation altogether.

Reviewer: ```suggestion
If a resource is missing a schema, Crossplane skips schema aware and integrity 
checking.

Crossplane only provides "missing schema" warnings or errors.
```

Author: Nope, just schema aware validation is skipped, not integrity checks. See the full logic below 👇 read "valid per se" as integrity check
 ![image](https://github.com/crossplane/docs/assets/5697904/c25949dd-723a-4556-a451-75dc82bbae93)

Reviewer: updated

Author: Not sure I get what you mean with the last sentence in the updated version.

The first sentence is an ok rewriting of the original version, although a bit oversimplified. I would be happier keeping the original version tbh

Reviewer: The last sentence is to be explicit about what is still being done. It's implied when you say. "skips schema aware" but this adds in meaning of "it still runs the other two checks"

Author: `Crossplane only provides "missing schema" and "integrity checking" warnings or errors.` 

The wording and the fact makes it sound like if Crossplane always only provided these two things.

A more correct version would be:
```suggestion
If a resource schema is missing, Crossplane skips schema aware validation altogether, but will still return an error in case of any integrity error and either a warning or an error depending on the mode selected for the missing schemas.
```
But I get Vale would probably not like it and it's a long sentence, which are the reasons why I tried avoiding getting in so much details.

Reviewer: I like that. Slight edit for our angry overlord Vale:

> If a resource schema is missing, Crossplane skips schema aware validation 
> but still returns an error for integrity errors and a warning or an error 
> for the missing schemas.

{{< /hint >}}

The following modes are available:

{{< table "table table-sm table-striped" >}}
| Mode     | Missing Schema | Schema Aware Error | Integrity Error |
| -------- | -------------- |--------------------|-----------------|
| `warn`   | Warning        | Warning            | Error           |
| `loose`  | Warning        | Error              | Error           |
| `strict` | Error          | Error              | Error           |
{{< /table >}}

Change the validation mode for a Composition with the
{{<hover label="mode" line="5">}}crossplane.io/composition-schema-aware-validation-mode{{</hover>}} 
annotation.

If not specified, the default mode is `warn`.

For example, to enable `loose` mode checking set the annotation value to 
{{<hover label="mode" line="5">}}loose{{</hover>}}.

```yaml {copy-lines="none",label="mode"}
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  annotations:
    crossplane.io/composition-schema-aware-validation-mode: loose
  # Removed for brevity
spec:
  # Removed for brevity
```

{{<hint "important" >}}

> Modes also apply to Compositions defined by Configuration packages, look at
> Crossplane's logs for warnings in case of issues. The default is just to warn,
> so unless the Configuration explicitly set a different mode, Crossplane creates
> the Composition even in case of errors, but you should still check warnings as
> they can surface bugs in the Configuration or in the schema aware validation
> logic, and either the author of the Configuration or the Crossplane team should
> fix it, respectively.

Reviewer: ```suggestion
Validation modes also apply to Compositions defined by Configuration packages.

View the Crossplane pod's logs with `kubectl logs` to see warnings or errors. 

By default Composition validation only issues warnings unless the 
Configuration explicitly sets a different mode. 

Crossplane creates the Composition from a Configuration package even if there 
are errors.
```

Author: Splitting things up this way make it slightly incorrect.

You have to view logs to see warnings only for compositions applied by Crossplane itself, splitting it up makes it look like for all warnings you have to look at the logs, which is false.

The default mode is to only emit warnings for schema aware validation errors, not for all kinds of validation, and unless the Composition defined in the Configuration set a different mode, the default will apply. So, compositions will be successfully applied, but warnings may be printed in the logs. While from your sentence it looks like when creating compositions for packages we always ignore errors.

Reviewer: Updated copy.

Author: `Unlike applying a Composition, Crossplane doesn’t reject errors in a Composition inside of a Configuration.` is not correct, Crossplane doesn't distinguish between Compositions coming from a user and the ones coming from a Configuration, they all hit the validating webhook and are all validated according to the same above logic, so if the mode set is something else than `warn`, a Composition coming from a Configuration could result in an error.

Reviewer: Sorry, I feel like I'm going in a circle here, I apologize.

So if I apply a configuration that has a Composition with `crossplane.io/composition-schema-aware-validation-mode: strict` what happens?

Does it apply the Configuration, reject the Composition and the package doesn't become `HEALTHY`?

Author: Yes, exactly.

Reviewer: one more rewrite. Let me know if this works.

Author: Not really 😅
A more correct, less vale-pleasing version would be:
```
Crossplane applies schema aware validation to Compositions defined by Configuration packages as well.

Depending on the validation mode set on the Composition, in case of schema aware validation issues,
Crossplane will log any warning to  its pod stdout or, if the Composition is fully rejected, mark the parent Configuration as unhealthy and show the error as condition's status.
```

However what I cared about is letting people know to look at logs to see warning and report them back if needed, they will notice if a composition is unhealthy no need to tell that.

are valid. 
<!-- vale write-good.TooWordy = YES -->

> Crossplane performs compositions "schema validation" by default. However, if you
> prefer to bypass this functionality, you can set the
> `--enable-composition-webhook-schema-validation=false` flag on the Crossplane
> pod.

Reviewer: ```suggestion
Crossplane performs compositions "schema validation" by default.  
Disable validation with the 
`--enable-composition-webhook-schema-validation=false` flag on the Crossplane
pod.
```

Reviewer: Update to pass Vale

Author: changed, have a look at it, I managed to make vale happy 🎉
